src/qp/packing_utils.py

```diff
@@ -0,0 +1,139 @@
+"""Integer packing utilities for qp"""
+
+import enum
+import numpy as np
+
+
+
+class PackingType(enum.Enum):
+
+    linear_from_rowmax = 0
+    log_from_rowmax = 1
+
+
+def linear_pack_from_rowmax(input_array):
+    """Pack an array into 8bit unsigned integers, using the maximum of each row as a refrence
+
+    This packs the values onto a linear grid for each row, running from 0 to row_max
+
+    Parameters
+    ----------
+    input_array : array_like
+        The values we are packing
+
+    Returns
+    -------
+    packed_array : array_like
+        The packed values
+    row_max : array_like
+        The max for each row, need to unpack the array
+    """
+    row_max = np.expand_dims(input_array.max(axis=1), -1)
+    return np.round(255 * input_array / row_max).astype(np.uint8), row_max
+
+
+def linear_unpack_from_rowmax(packed_array, row_max):
+    """Unpack an array into 8bit unsigned integers, using the maximum of each row as a refrence
+
+    Parameters
+    ----------
+    packed_array : array_like
+        The packed values
+    row_max : array_like
+        The max for each row, need to unpack the array
+
+
+    Returns
+    -------
+    unpacked_array : array_like
+        The unpacked values
+    """
+    unpacked_array = row_max * packed_array / 255.
+    return unpacked_array
+
+
+def log_pack_from_rowmax(input_array, log_floor=-3.):
+    """Pack an array into 8bit unsigned integers, using the maximum of each row as a refrence
+
+    This packs the values onto a log grid for each row, running from row_max / 10**log_floor to row_max
+
+    Parameters
+    ----------
+    input_array : array_like
+        The values we are packing
+    log_floor: float
+        The logarithmic floor used for the packing
+
+    Returns
+    -------
+    packed_array : array_like
+        The packed values
+    row_max : array_like
+        The max for each row, need to unpack the array
+    """
+    neg_log_floor = -1. * log_floor
```

Comment:
it is actually 'negative_log_floor',  the exponentiation (10** something) comes in the unpacking, the factor of -1 here is just to make the intermediate mapping run from 0 to 1 (so that we can then multiply by 255 to map it onto 0 to 255)